src/vmm/src/devices/virtio/net/rx_buffer.rs

```diff
@@ -0,0 +1,510 @@
+    }
+
+    /// Writes bytes from a slice into buffer.
+    pub fn write(&mut self, mut bytes: &[u8]) -> Result<(), RxBufferError> {
+        for iov in self.iovecs.as_mut_slice() {
+            if bytes.is_empty() {
+                break;
+            }
+            let iov_slice_len = bytes.len().min(iov.iov_len);
+            // SAFETY: The user space pointer and the length were verified during
+            // the iovec creation.
+            let slice =
+                unsafe { std::slice::from_raw_parts_mut(iov.iov_base.cast(), iov_slice_len) };
+            slice.copy_from_slice(&bytes[0..slice.len()]);
+            bytes = &bytes[slice.len()..];
+        }
+        if !bytes.is_empty() {
+            Err(RxBufferError::WriteOverflow)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Finish packet processing by removing used iovecs from the buffer and
+    /// writing information about used descriptor chains into the queue.
+    ///
+    /// # Safety
+    /// `RxBuffer` should not be empty when this method is called because it
+    /// assumes there is at least one chain holding data.
+    pub unsafe fn finish_packet(&mut self, mut bytes_written: u32, rx_queue: &mut Queue) {
+        // This function is called only after some bytes were written to the
+        // buffer. This means the iov_ring cannot be empty.
+        debug_assert!(!self.iovecs.is_empty());
+        let packet_head = self.iovecs.as_slice()[0].iov_base;
+        let mut used_heads = 0;
+
+        let mut write_used = |head_index: u16, bytes_written: u32, rx_queue: &mut Queue| {
+            if let Err(err) = rx_queue.write_used_element(
+                (rx_queue.next_used + Wrapping(self.used_descriptors)).0,
+                head_index,
+                bytes_written,
+            ) {
+                error!(
+                    "net: Failed to add used descriptor {} of length {} to RX queue: {err}",
+                    head_index, bytes_written
+                );
+            }
+            used_heads += 1;
+            self.used_descriptors += 1;
+        };
+
+        loop {
+            let iov_info = self
+                .chain_infos
+                .pop_front()
+                .expect("This should never happen if write to the buffer succeded.");
+            self.iovecs.pop_front(usize::from(iov_info.chain_len));
+            self.total_capacity -= iov_info.chain_capacity;
+
+            if bytes_written <= iov_info.chain_capacity {
+                write_used(iov_info.head_index, bytes_written, rx_queue);
+                break;
+            } else {
+                write_used(iov_info.head_index, iov_info.chain_capacity, rx_queue);
+                bytes_written -= iov_info.chain_capacity;
+            }
+        }
+        // SAFETY: The user space pointer was verified at the  point of creation.
+        #[allow(clippy::transmute_ptr_to_ref)]
+        let header: &mut virtio_net_hdr_v1 = unsafe { std::mem::transmute(packet_head) };
```

Comment:
I don't think we should check for alignment here. If the buffer is not aligned, then the guest just creates more problem for itself as it will need to check alignment of it's own buffer. Also I cannot find alignment checks in the vhost-net code in the kernel.

Reply:
If the reference we create is unaligned, we are triggering undefined behavior, see the second point in https://doc.rust-lang.org/reference/behavior-considered-undefined.html. We cannot trust the guest that it will "not cause problems for itself", and rustc will most definitely assume the references are aligned and thus not emit unaligned load instructions for performance reasons.

But even without that, we cannot create a reference _at all_ here, because the packet lives in guest memory

Reply:
updated the way we write to the `num_buffers` and added checks for alignment and buffer size.